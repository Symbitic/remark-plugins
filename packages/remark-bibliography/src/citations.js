import chicago from 'style-chicago'
import CSL from 'citeproc'
import enUS from 'locale-en-us'
import visit from 'unist-util-visit'

export default function citations (tree, items, opts = {}) {
  const style = opts.style || chicago
  const locale = opts.locale || enUS
  
  const sys = {
    retrieveLocale: () => locale,
    retrieveItem: id => items[id]
  }
  
  // Create Citeproc and add items
  const citeproc = new CSL.Engine(sys, style)
  citeproc.updateItems(Object.keys(items))
  
  // Create citations
  const citations = Object.entries(items)
    .map(([ key, val ]) => ([
      key,
      citeproc.makeCitationCluster([ val ])
        // .replace(/^\(.+)/)$/, '$1')
    ]))
    .reduce((acc, val) => Object.assign(acc, {
      [val[0]]: val[1]
    }), {})
  
  // Create bibliography
  const bibliography = citeproc.makeBibliography()
  const ids = bibliography[0].entry_ids.map(ids => ids.join(''))
  const bib = bibliography[1]
    .map(str => str
      .trim()
      .replace(/<div[^>]+>/g, '')
      .replace(/<\/div>/g, '')
      .replace(/<\/i>/g, '<i>')
      .split('<i>')
      .map((item, i) => (i % 2 === 0 ? {
        type: 'text',
        value: item
      } : {
        type: 'emphasis',
        children: [{ type: 'text', value: item }]
      }))
    )

  // Add citations to MDAST
  Object.entries(citations).forEach(([ id, cite ]) => {
    const key = `(@${id})`
    visit(tree, [ 'text' ], (node, i, parent) => {
      const contains = node.value.includes(key)
      if (!contains) {
        return
      }
      const values = node.value.split(key)
      const children = [
        {
          type: 'text',
          value: values[0]
        },
        {
          type: 'link',
          url: `#ref-${id}`,
          data: {
            hProperties: {
              className: 'citation'
            }
          },
          children: [{ type: 'text', value: cite }]
        },
        {
          type: 'text',
          value: values[1]
        }
      ]
      parent.children.splice(i, 1, ...children)
    })
  })
  
  // Append bibliography to MDAST
  tree.children = tree.children.concat({
    type: 'element',
    data: {
      hName: 'div',
      hProperties: {
        className: 'references'
      }
    },
    children: [
      {
        type: 'heading',
        depth: 2,
        children: [
          {
            type: 'text',
            value: 'References'
          }
        ]
      },
      {
        type: 'list',
        ordered: true,
        spread: false,
        data: {
          hProperties: { className: 'ref-list' }
        },
        children: bib.map((entry, i) => ({
          type: 'listItem',
          spread: false,
          data: {
            hProperties: { id: `ref-${ids[i]}`, className: 'reference' }
          },
          children: entry
        }))
      }
    ]
  })
  
  return tree
}
