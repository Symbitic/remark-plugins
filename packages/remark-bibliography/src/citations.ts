import chicago from 'style-chicago'
import CSL from 'citeproc'
import enUS from 'locale-en-us'
import { visit } from 'unist-util-visit'
import type { Data, Literal, Node, Parent } from 'unist'

export interface Options {
  style: Record<string, any>;
  locale: Record<string, any>;
}

export default function citations(root: Node<Data>, items: any, opts: Partial<Options> = {}) {
  const tree = root as Parent<Node<Data>, Data>;
  const style = opts.style || chicago
  const locale = opts.locale || enUS

  const sys = {
    retrieveLocale: () => locale,
    retrieveItem: (id: string) => items[id]
  }

  // Create Citeproc and add items
  const citeproc = new CSL.Engine(sys, style)
  citeproc.updateItems(Object.keys(items))

  // Create citations
  const citations = Object.entries(items)
    .map(([key, val]) => ([
      key,
      citeproc.makeCitationCluster([val])
        // .replace(/^\(.+)/)$/, '$1')
    ]))
    .reduce((acc, val) => Object.assign(acc, {
      [val[0]]: val[1]
    }), {})

  // Create bibliography
  const bibliography = citeproc.makeBibliography()
  const ids = bibliography[0].entry_ids.map((ids: string[]) => ids.join(''))
  const bib = bibliography[1]
    .map((str: string) => str
      .trim()
      .replace(/<div[^>]+>/g, '')
      .replace(/<\/div>/g, '')
      .replace(/<\/i>/g, '<i>')
      .split('<i>')
      .map((item, i) => (i % 2 === 0 ? {
        type: 'text',
        value: item
      } : {
        type: 'emphasis',
        children: [{ type: 'text', value: item }]
      }))
    )

  // Add citations to MDAST
  Object.entries(citations).forEach(([id, cite]) => {
    const key = `(@${id})`
    visit(tree, ['text'], (node, i, parent) => {
      const { value } = node as Literal<string>

      const contains = value.includes(key)
      if (!contains) {
        return
      }
      const values = value.split(key)
      const children = [
        {
          type: 'text',
          value: values[0]
        },
        {
          type: 'link',
          url: `#ref-${id}`,
          data: {
            hProperties: {
              className: 'citation'
            }
          },
          children: [{ type: 'text', value: cite }]
        },
        {
          type: 'text',
          value: values[1]
        }
      ]
      parent!.children.splice(i!, 1, ...children)
    })
  })

  // Append bibliography to MDAST
  tree.children = tree.children.concat({
    type: 'element',
    data: {
      hName: 'div',
      hProperties: {
        className: 'references'
      }
    },
    children: [
      {
        type: 'heading',
        depth: 2,
        children: [
          {
            type: 'text',
            value: 'References'
          }
        ]
      },
      {
        type: 'list',
        ordered: true,
        spread: false,
        data: {
          hProperties: { className: 'ref-list' }
        },
        children: bib.map((entry, i) => ({
          type: 'listItem',
          spread: false,
          data: {
            hProperties: { id: `ref-${ids[i]}`, className: 'reference' }
          },
          children: entry
        }))
      }
    ]
  } as any)

  return tree as Node<Data>
}
