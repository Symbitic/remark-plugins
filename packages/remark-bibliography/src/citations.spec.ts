/* eslint-env jest */
test('citations', () => {
  expect(2+2).toBe(4);
})
/*
import citations from './citations'

type Case = [ string, any, Record<string, any>, any? ]

const case1: Case = [
  'replace a citation correctly',
  {
    type: 'root',
    children: [{
      type: 'paragraph',
      children: [{ type: 'text', value: 'This is a citation: (@singh2016).' }]
    }]
  },
  {
    gregory2014: {
      type: 'book',
      title: 'Game Engine Architecture, Second Edition.',
      author: [{ given: 'Jason', family: 'Gregory' }],
      publisher: 'CRC Press',
      issued: { 'date-parts': [ [ 2014 ] ] },
      URL: 'http://www.gameenginebook.com/',
      DOI: '10.1186/1758-2946-3-47',
      id: 'gregory2014'
    },
    singh2016: {
      type: 'book',
      title: 'Vulkan Essentials',
      author: [{ given: 'Parminder', family: 'Singh' }],
      publisher: 'Packt Publishing',
      issued: { 'date-parts': [ [ 2016 ] ] },
      URL: 'https://www.packtpub.com/application-development/vulkan-essentials',
      id: 'singh2016'
    },
    moller2008: {
      id: 'moller2008',
      type: 'book',
      title: 'Real Time Rendering, Third Edition',
      author: [
        { given: 'Thomas', family: 'Akenine-Moller' },
        { given: 'Eric', family: 'Haines' },
        { given: 'Naty', family: 'Hoffman' }
      ],
      publisher: 'CRC Press',
      issued: { 'date-parts': [ [ 2008 ] ] },
      URL: 'http://www.realtimerendering.com/'
    }
  }
]

const fixtures = [
  case1
]

describe.each(fixtures)('citations', (name, tree, items, opts) => {
  it(`should ${name}`, () => {
    expect(citations(tree, items, opts)).toMatchSnapshot()
  })
})
*/
