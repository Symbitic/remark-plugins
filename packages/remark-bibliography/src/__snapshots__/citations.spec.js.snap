// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`citations should replace a citation correctly 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "This is a citation: ",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "(Singh 2016)",
            },
          ],
          "data": Object {
            "hProperties": Object {
              "className": "citation",
            },
          },
          "type": "link",
          "url": "#ref-singh2016",
        },
        Object {
          "type": "text",
          "value": ".",
        },
      ],
      "type": "paragraph",
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "References",
            },
          ],
          "depth": 2,
          "type": "heading",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Akenine-Moller, Thomas, Eric Haines, and Naty Hoffman. 2008. ",
                },
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "Real Time Rendering, Third Edition",
                    },
                  ],
                  "type": "emphasis",
                },
                Object {
                  "type": "text",
                  "value": ". CRC Press. http://www.realtimerendering.com/.",
                },
              ],
              "data": Object {
                "hProperties": Object {
                  "className": "reference",
                  "id": "ref-moller2008",
                },
              },
              "spread": false,
              "type": "listItem",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Gregory, Jason. 2014. ",
                },
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "Game Engine Architecture, Second Edition.",
                    },
                  ],
                  "type": "emphasis",
                },
                Object {
                  "type": "text",
                  "value": " CRC Press. https://doi.org/10.1186/1758-2946-3-47.",
                },
              ],
              "data": Object {
                "hProperties": Object {
                  "className": "reference",
                  "id": "ref-gregory2014",
                },
              },
              "spread": false,
              "type": "listItem",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Singh, Parminder. 2016. ",
                },
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "Vulkan Essentials",
                    },
                  ],
                  "type": "emphasis",
                },
                Object {
                  "type": "text",
                  "value": ". Packt Publishing. https://www.packtpub.com/application-development/vulkan-essentials.",
                },
              ],
              "data": Object {
                "hProperties": Object {
                  "className": "reference",
                  "id": "ref-singh2016",
                },
              },
              "spread": false,
              "type": "listItem",
            },
          ],
          "data": Object {
            "hProperties": Object {
              "className": "ref-list",
            },
          },
          "ordered": true,
          "spread": false,
          "type": "list",
        },
      ],
      "data": Object {
        "hName": "div",
        "hProperties": Object {
          "className": "references",
        },
      },
      "type": "element",
    },
  ],
  "type": "root",
}
`;
